### PR TITLE
Add Sky Drop to Battle Engine

### PR DESCRIPTION
**Pull request by WiserVisor**
Added in the move Sky Drop, which came with some new macros, methods, and memory allocations to get it to work.

## Description
Sky Drop works just as it does from Gen 6 onwards.

Issues/Glitches

When a Pokemon is released from Sky Drop, the text: "{TARGET} was released from Sky Drop!" doesn't appear. This is less of an issue and more the fact that I found that bit of text unnecessary in the games.

If a Pokemon that used Sky Drop is in the middle of the semi-invulnerable turn and falls asleep due to Yawn, the attacker's sprite will appear as normal, but the target's sprite will still be invisible. Both Pokemon are still properly taken out of the semi-invulnerable state.

## **Discord contact info**
N/A
Contact me on PokeCommunity
WiserVisor https://www.pokecommunity.com/member.php?u=682852

### CODE REVIEW

File: include/battle.h [hunk 1]
```diff
@@ -542,6 +542,7 @@ struct BattleStruct
     u8 sameMoveTurns[MAX_BATTLERS_COUNT]; // For Metronome, number of times the same moves has been SUCCESFULLY used.
     u16 moveEffect2; // For Knock Off
     u16 changedSpecies[PARTY_SIZE]; // For Zygarde or future forms when multiple mons can change into the same pokemon.
+	u8 skyDropTargets[MAX_BATTLERS_COUNT]; // For Sky Drop, to account for if multiple Pokemon use Sky Drop in a double battle.
```
Comment (LOuroboros) [inline include/battle.h:545]: Use empty spaces instead of tabs.

File: include/constants/battle.h [hunk 2]
```diff
@@ -171,6 +171,7 @@
 #define STATUS3_LASER_FOCUS             (1 << 29)
 #define STATUS3_ELECTRIFIED             (1 << 30)
 #define STATUS3_POWER_TRICK             (1 << 31)
+#define STATUS3_SKY_DROPPED				(STATUS3_ON_AIR | STATUS3_UNDERGROUND)
```
Comment (LOuroboros) [inline include/constants/battle.h:174]: The value of the `STATUS3_SKY_DROPPED` constant is not properly aligned with the values of the constants above and below it.

File: include/constants/battle.h [hunk 3]
```diff
@@ -342,7 +343,8 @@
 #define MOVE_EFFECT_THROAT_CHOP         0x43
 #define MOVE_EFFECT_INCINERATE          0x44
 #define MOVE_EFFECT_BUG_BITE            0x45
-#define NUM_MOVE_EFFECTS                0x46
+#define MOVE_EFFECT_SKY_DROP			0x46
```
Comment (LOuroboros) [inline include/constants/battle.h:346]: The value of the `MOVE_EFFECT_SKY_DROP` constant is not properly aligned with the values of the constants above and below it.

File: include/constants/battle_script_commands.h [hunk 4]
```diff
@@ -168,6 +168,9 @@
 #define VARIOUS_JUMP_IF_ABSENT                  101
 #define VARIOUS_DESTROY_ABILITY_POPUP           102
 #define VARIOUS_TOTEM_BOOST                     103
+#define VARIOUS_JUMP_IF_UNDER_200				104
+#define VARIOUS_SET_SKY_DROP					105
+#define VARIOUS_CLEAR_SKY_DROP					106
```
Comment (LOuroboros) [inline include/constants/battle_script_commands.h:173]: The values of these constants are not properly aligned with the values of the constants above and below them.

File: src/battle_ai_switch_items.c [hunk 5]
```diff
@@ -792,7 +792,9 @@ static bool8 ShouldUseItem(void)
     u8 validMons = 0;
     bool8 shouldUse = FALSE;
 
-    if (gBattleTypeFlags & BATTLE_TYPE_INGAME_PARTNER && GetBattlerPosition(gActiveBattler) == B_POSITION_PLAYER_RIGHT)
+    if ((gBattleTypeFlags & BATTLE_TYPE_INGAME_PARTNER && GetBattlerPosition(gActiveBattler) == B_POSITION_PLAYER_RIGHT)
+        // If Pokemon is currently being held by Sky Drop
```
Comment (LOuroboros) [inline src/battle_ai_switch_items.c:796]: The comment gets in the way if you put it in the middle of an ongoing check. I suggest to put it above the `if ((gBattleTypeFlags & BATTLE_TYPE_INGAME_PARTNER && GetBattlerPosition(gActiveBattler) == B_POSITION_PLAYER_RIGHT)` instead.

File: src/battle_main.c [hunk 6]
```diff
@@ -3842,10 +3842,12 @@ static void HandleTurnActionSelectionState(void)
                     }
                     break;
                 case B_ACTION_USE_ITEM:
-                    if (gBattleTypeFlags & (BATTLE_TYPE_LINK
+                    if ((gBattleTypeFlags & (BATTLE_TYPE_LINK
                                             | BATTLE_TYPE_FRONTIER_NO_PYRAMID
                                             | BATTLE_TYPE_EREADER_TRAINER
                                             | BATTLE_TYPE_RECORDED_LINK))
+						// Or if currently held by Sky Drop
+						|| ((gStatuses3[gActiveBattler] & STATUS3_ON_AIR) && (gStatuses3[gActiveBattler] & STATUS3_UNDERGROUND)))
```
Comment (LOuroboros) [inline src/battle_main.c:3850]: Use empty spaces instead of tabs so the conditions you added to the check are properly aligned with the ones that were already there.

File: src/battle_message.c [hunk 7]
```diff
@@ -1245,6 +1247,8 @@ const u8 *const gBattleStringsTable[BATTLESTRINGS_COUNT] =
     [STRINGID_AURABREAKENTERS - 12] = sText_AuraBreakActivates,
     [STRINGID_COMATOSEENTERS - 12] = sText_ComatoseActivates,
     [STRINGID_SCREENCLEANERENTERS - 12] = sText_ScreenCleanerActivates,
+	[STRINGID_PKMNTOOKTARGETHIGH - 12] = sText_PkmnTookTargetHigh,
+	[STRINGID_TARGETTOOHEAVY - 12] = sText_TargetTooHeavy,
```
Comment (LOuroboros) [inline src/battle_message.c:1251]: Replace those tabs with empty spaces.

File: src/battle_message.c [hunk 8]
```diff
@@ -1400,6 +1404,7 @@ const u16 gFirstTurnOfTwoStringIds[] =
     STRINGID_VANISHEDINSTANTLY,         // MOVE_PHANTOM_FORCE
     STRINGID_PKNMABSORBINGPOWER,        // MOVE_GEOMANCY
     STRINGID_CLOAKEDINAFREEZINGLIGHT,   // MOVE_FREEZE_SHOCK
+	STRINGID_PKMNTOOKTARGETHIGH, //MOVE_SKY_DROP
```
Comment (LOuroboros) [inline src/battle_message.c:1407]: Replace that tab with 4 empty spaces and add some spaces inbetween the last comma and the comment in order to align it properly with the ones in the lines above it.

File: src/battle_script_commands.c [hunk 9]
```diff
@@ -8335,6 +8335,40 @@ static void Cmd_various(void)
...
+		}
+		break;
+	case VARIOUS_CLEAR_SKY_DROP:
+		gStatuses3[gBattlerTarget] &= ~STATUS3_SKY_DROPPED;
+		gBattleMons[gBattlerTarget].status2 &= ~STATUS2_ESCAPE_PREVENTION;
+		if (gBattleStruct->skyDropTargets[0] - 4 == gBattlerAttacker)
+			gBattleStruct->skyDropTargets[0] = 0;
+		else
+			gBattleStruct->skyDropTargets[2] = 0;
+		break;
```
Comment (LOuroboros) [inline src/battle_script_commands.c:8371]: Replace every single tab here with 4 empty spaces each.

File: src/battle_script_commands.c [hunk 10]
```diff
@@ -10906,6 +10940,7 @@ static void Cmd_setsemiinvulnerablebit(void)
     {
     case MOVE_FLY:
     case MOVE_BOUNCE:
+	case MOVE_SKY_DROP:
```
Comment (LOuroboros) [inline src/battle_script_commands.c:10943]: Replace the tab here with 4 empty spaces.

File: src/battle_util.c [hunk 11]
```diff
@@ -1189,9 +1189,24 @@ void CancelMultiTurnMoves(u8 battler)
...
+		gStatuses3[gBattleStruct->skyDropTargets[1]] &= ~STATUS3_SKY_DROPPED;
+		gBattleMons[gBattleStruct->skyDropTargets[1]].status2 &= ~STATUS2_ESCAPE_PREVENTION;
+		gBattleStruct->skyDropTargets[0] = 0;
+	}
+	else if (gBattleStruct->skyDropTargets[2] - 4 == battler)
+	{
+		gStatuses3[gBattleStruct->skyDropTargets[3]] &= ~STATUS3_SKY_DROPPED;
+		gBattleMons[gBattleStruct->skyDropTargets[3]].status2 &= ~STATUS2_ESCAPE_PREVENTION;
+		gBattleStruct->skyDropTargets[2] = 0;
+	}
```
Comment (LOuroboros) [inline src/battle_util.c:1205]: Replace every single tab here with 4 empty spaces each.

File: src/battle_util.c [hunk 12]
```diff
@@ -2896,6 +2912,16 @@ u8 AtkCanceller_UnableToUseMove(void)
...
+	case CANCELLER_SKY_DROP:
+			// If Pokemon has STATUS3_SKY_DROPPED
+			if((gStatuses3[gBattlerAttacker] & STATUS3_ON_AIR) && (gStatuses3[gBattlerAttacker] & STATUS3_UNDERGROUND))
+			{
+				gBattlescriptCurrInstr = BattleScript_MoveEnd;
+                		gHitMarker |= HITMARKER_UNABLE_TO_USE_MOVE;
+                		effect = 1;
+			}
+			gBattleStruct->atkCancellerTracker++;
+            break;
```
Comment (LOuroboros) [inline src/battle_util.c:2924]: Replace every single tab here with 4 empty spaces each.

File: src/battle_util.c [hunk 13]
```diff
@@ -4385,7 +4411,9 @@ u8 AbilityBattleEffects(u8 caseID, u8 battler, u16 ability, u8 special, u16 move
              && (gMultiHitCounter == 0 || gMultiHitCounter == 1)
              && !(GetBattlerAbility(gBattlerAttacker) == ABILITY_SHEER_FORCE && gBattleMoves[gCurrentMove].flags & FLAG_SHEER_FORCE_BOOST)
              && (CanBattlerSwitch(battler) || !(gBattleTypeFlags & BATTLE_TYPE_TRAINER))
-             && !(gBattleTypeFlags & BATTLE_TYPE_ARENA))
+             && !(gBattleTypeFlags & BATTLE_TYPE_ARENA)
+	     // Not currently held by Sky Drop
+	     && !(gStatuses3[battler] & STATUS3_ON_AIR && gStatuses3[battler] & STATUS3_UNDERGROUND))
```
Comment (LOuroboros) [inline src/battle_util.c:4416]: Replace the tabs here with 4 empty spaces each.

**Review by LOuroboros — CHANGES_REQUESTED** [review]
On top of the syntax related corrections, you should remove the github template from this PR since it's completely unrelated to the addition of Sky Drop's effect.

File: include/constants/battle.h (hunk 2 above)
Comment (LOuroboros) [reply include/constants/battle.h:174]: Also, you're using tabs when you should be using empty spaces.

File: include/constants/battle.h (hunk 3 above)
Comment (LOuroboros) [reply include/constants/battle.h:346]: Also, you're using tabs when you should be using empty spaces.

File: include/constants/battle_script_commands.h (hunk 4 above)
Comment (LOuroboros) [reply include/constants/battle_script_commands.h:173]: Also, you're using tabs when you should be using empty spaces.

File: src/battle_main.c [hunk 14]
```diff
@@ -3842,10 +3842,12 @@ static void HandleTurnActionSelectionState(void)
                     }
                     break;
                 case B_ACTION_USE_ITEM:
-                    if (gBattleTypeFlags & (BATTLE_TYPE_LINK
+                    if ((gBattleTypeFlags & (BATTLE_TYPE_LINK
                                             | BATTLE_TYPE_FRONTIER_NO_PYRAMID
                                             | BATTLE_TYPE_EREADER_TRAINER
                                             | BATTLE_TYPE_RECORDED_LINK))
+					    // Or if currently held by Sky Drop
+					    || ((gStatuses3[gActiveBattler] & STATUS3_ON_AIR) && (gStatuses3[gActiveBattler] & STATUS3_UNDERGROUND)))
```
Comment (LOuroboros) [inline src/battle_main.c:3850]: Spaces instead of tabs.

Comment (WiserVisor) [reply src/battle_main.c]: I've changed the tabs to spaces, but for some reason, it still looks like there are tabs from the pull request window.

File: src/battle_util.c [hunk 15]
```diff
@@ -2896,6 +2912,16 @@ u8 AtkCanceller_UnableToUseMove(void)
             gStatuses3[gBattlerAttacker] &= ~(STATUS3_GRUDGE);
             gBattleStruct->atkCancellerTracker++;
             break;
+        case CANCELLER_SKY_DROP:
+            // If Pokemon has STATUS3_SKY_DROPPED
+            if((gStatuses3[gBattlerAttacker] & STATUS3_ON_AIR) && (gStatuses3[gBattlerAttacker] & STATUS3_UNDERGROUND))
```
Comment (LOuroboros) [inline src/battle_util.c:2917]: Throw in an empty space inbetween `if` and the first parenthesis.

File: src/battle_script_commands.c [hunk 16]
```diff
@@ -10960,6 +10994,7 @@ static void Cmd_setsemiinvulnerablebit(void)
     {
     case MOVE_FLY:
     case MOVE_BOUNCE:
+    case MOVE_SKY_DROP:
         gStatuses3[gBattlerAttacker] |= STATUS3_ON_AIR;
```
Comment (ExpoSeed) [inline src/battle_script_commands.c:10998]: set the on air bit on the target as well, because the target is also in air

**Review by ExpoSeed — CHANGES_REQUESTED** [review]

File: .github/pull_request_template.md [hunk 17]
```diff
@@ -5,4 +5,4 @@
 
 ## **Discord contact info**
 <!--- formatted as name#numbers, e.g. PikalaxALT#5823 -->
-<!--- Contributors must join https://discord.gg/d5dubZ3 -->
\ No newline at end of file
+<!--- Contributors must join https://discord.gg/d5dubZ3 -->
```
Comment (ExpoSeed) [inline .github/pull_request_template.md:8]: don't put unrelated minor changes in the pr

File: src/battle_script_commands.c (hunk 16 above)
Comment (WiserVisor) [reply src/battle_script_commands.c:10998]: This is handled within Cmd_various function in this file, with the VARIOUS_SET_SKY_DROP and VARIOUS_CLEAR_SKY_DROP cases, as there are a few other things that need to be handled when the target is affected by Sky Drop. I also feel that it is unwise to add code that pertains to only one particular event to a function that is used for multiple other events, at least when it can be reasonably avoided, which it can in this situation.

Comment (ExpoSeed) [reply src/battle_script_commands.c:10998]: ah you're right, i didn't realize that, my bad

File: .github/pull_request_template.md (hunk 17 above)
Comment (WiserVisor) [reply .github/pull_request_template.md:8]: I do not know how to take this change out of the pr. When I analyze the two lines, they seem to be identical, so I don't see why there's a change here in the first place.

Comment (ExpoSeed) [reply .github/pull_request_template.md:8]: it looks like a space at the end was deleted i think

Comment (WiserVisor) [reply .github/pull_request_template.md:8]: I added a space, and the space was marked as a change.

**Comment by ultima-soul** [comment]
Could you please remake this PR with only commits that implement the necessary changes? And include the suggested corrections you had to do as well. The commit history is too messed up in this PR.

Thanks in advance for your cooperation @WiserVisor!
